pkg/placement/placement-resolver.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// A PlacementResolver holds a collection of placement resolutions.
+// The collection is indexed by placementKey string, the resolver does not
+// care what the strings are. The resolution for a given key can be updated,
+// exported and compared to the PlacementDecision representation.
+// All functions in this interface are thread-safe, and no method parameter
+// gets mutated.
```

Comment:
By anyone, ever? Unlikely to be true. Slices and sets are constructed incrementally, thus mutated. Maybe you want to say something like the following.

> Nothing mutates any method parameter after it is passed in a call to one of these methods.

Reply:
> Nothing mutates any method-parameter during a call to one of these methods.